### ISSUE TITLE
FrameRecorder FPS & WebM Recording Issue

### ISSUE THREAD

**Issue by MarcWoodyard** [bytedeco/javacv#1481]
I have a couple of issues with the code below.

1.) When it captures frames and records, the output video is too fast. I looked it up online for a couple of hours and the only thing I found was that the FPS on the frame grabber and the frame recorder didn't match. So I set them both to 20 FPS, and the issue improved, but when I recorded a two-minute video stream, it records all two minutes, but the output is only about one minute and thirty seconds (the video looks like it's being played in fast forward).

2.) The code below won't record to a .webm file format, I've only been able to record to a .mp4 format. It will only create an empty file that's 0KB in size and won't record any data to it.

```java
public class JavaCVStreamRecorder extends Thread {

    private static DateTimeFormatter fileFormatter = DateTimeFormatter.ofPattern("MM-dd-yyyy hh-mm-ss", Locale.ENGLISH);
    private static DateTimeFormatter directoryFormatter = DateTimeFormatter.ofPattern("MM-dd-yyyy", Locale.ENGLISH);
    private IPCameraFrameGrabber frameGrabber;
    private FrameRecorder frameRecorder;
    private CameraScreenshot cameraScreenshot = new CameraScreenshot();
    private Thread subThread;
    private String cameraName;
    private String streamURL;
    private int width;
    private int height;
    private LocalDateTime start;
    private LocalDateTime end;
    private boolean recordingEnabled;
    private boolean grabFrames = true;
    private int fps = 20;

    public JavaCVStreamRecorder(int id, String cameraName, String streamURL, int width, int height) {
        this.cameraName = cameraName;
        this.streamURL = streamURL;
        this.width = width;
        this.height = height;
        this.recordingEnabled = CameraManager.getCamera(id).getRecordingEnabled();
    }

    @Override
    public final void start() {
        try {
            this.frameGrabber = new IPCameraFrameGrabber(this.streamURL, 3, 3, TimeUnit.SECONDS);

            if (streamURL.contains("rtsp")) {
                this.frameGrabber.setFormat("rtsp");
                this.frameGrabber.setOption("rtsp_transport", "tcp");
            } else if (streamURL.contains("mjpg"))
                this.frameGrabber.setFormat("mjpeg");

            // Enable hardware acceleration (https://github.com/bytedeco/javacpp-presets/issues/708)
            this.frameGrabber.setOption("hwaccel", "opencl");

            this.frameGrabber.setFrameRate(this.fps);

            this.frameGrabber.start();

            // We're setting the resolution after we start because we need to grab a frame so we can
            // detect the frame's resolution.

            // If width & height are not zero, then it means that we set a specific resolution
            // (video stills for web interface), if they are 0, then it means that we want the
            // full resolution.
            if (this.width != 0 || this.height != 0) {
                this.frameGrabber.setImageWidth(this.width);
                this.frameGrabber.setImageHeight(this.height);
            } else {
                Frame frame = this.frameGrabber.grabFrame();
                this.frameGrabber.setImageWidth(frame.imageWidth);
                this.frameGrabber.setImageHeight(frame.imageHeight);
            }

            if (this.recordingEnabled)
                this.createNewRecording();

            this.subThread = new Thread(() -> {
                Frame frame = null;
                long startTime = 0;
                long videoTS = 0;

                while (grabFrames) {
                    try {
                        if (this.recordingEnabled && LocalDateTime.now().isAfter(this.end)) {
                            this.frameRecorder.stop();
                            this.createNewRecording();
                        }

                        // Prevents Socket Closed exception if thread is .interrupt() is
                        // called while inside of the while loop.
                        try {
                            frame = this.frameGrabber.grab();
                        } catch (Exception e) {
                            if (!e.getMessage().contains("Socket closed"))
                                e.printStackTrace();
                        }

                        if (frame != null) {
                            this.cameraScreenshot.updateImage(frame);

                            if (this.recordingEnabled) {

                                if (startTime == 0)
                                    startTime = System.currentTimeMillis();


                                // Create timestamp for this frame
                                videoTS = 1000 * (System.currentTimeMillis() - startTime);

                                // Check for AV drift
                                if (videoTS > this.frameRecorder.getTimestamp()) {
                                    /*
                                    System.out.println(
                                            "Lip-flap correction: "
                                                    + videoTS + " : "
                                                    + this.frameRecorder.getTimestamp() + " -> "
                                                    + (videoTS - this.frameRecorder.getTimestamp()));
                                     */

                                    // We tell the recorder to write this frame at this timestamp
                                    this.frameRecorder.setTimestamp(videoTS);
                                }

                                this.frameRecorder.record(frame);
                                Thread.sleep((long) (1000 / this.fps));
                            }

                        }
                    } catch (NullPointerException ignored) {
                    } catch (Exception e) {
                        e.printStackTrace();
                    }
                }
            });

            this.subThread.start();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private void createNewRecording() {
        try {

            // TODO Change this to a timer. If time changes, it will create a large file.
            this.start = LocalDateTime.now();
            this.end = LocalDateTime.now().plusMinutes(2);

            File recordingDir;

            do {
                recordingDir = new File(ProgramSettings.getRecordingDirectory() +
                        File.separator +
                        this.cameraName +
                        File.separator +
                        directoryFormatter.format(this.start));

                if (!recordingDir.exists()) {
                    // If we can't create the recording dir specified in ProgramSettings(), then revert to the default location.
                    if (!recordingDir.mkdirs()) {
                        ProgramSettings.setRecordingDirectory(ProgramSettings.getDefaultRecordingDir());
                        ProgramSettings.save();
                    }
                }
            } while (!recordingDir.exists());

            String videoFileLocation = recordingDir.getAbsoluteFile() +
                    File.separator +
                    cameraName + " [" + fileFormatter.format(this.start) + "] - [" + fileFormatter.format(this.end) + "].webm";

            this.frameRecorder = new OpenCVFrameRecorder(
                    videoFileLocation,
                    this.frameGrabber.getImageWidth(),
                    this.frameGrabber.getImageHeight());

            this.frameRecorder.setInterleaved(true);

            // Decrease "startup" latency in FFMPEG (https://trac.ffmpeg.org/wiki/StreamingGuide)
            this.frameRecorder.setVideoOption("tune", "zerolatency");

            // Key frame interval, in our case every 2 seconds -> FPS * 2 = GOP length
            this.frameRecorder.setGopSize(this.fps * 2);

            // Tradeoff between quality and encode speed possible values are ultrafast, superfast, veryfast, faster, fast,
            // medium, slow, slower, veryslow, ultrafast offers us the least amount of compression (lower encoder CPU) at the
            // cost of a larger stream size at the other end, veryslow provides the best compression (high encoder CPU) while
            // lowering the stream size (https://trac.ffmpeg.org/wiki/Encode/H.264)
            this.frameRecorder.setVideoOption("preset", "ultrafast");

            // Constant Rate Factor (https://trac.ffmpeg.org/wiki/Encode/H.264)
            this.frameRecorder.setVideoOption("crf", "28");

            this.frameRecorder.setVideoCodec(avcodec.AV_CODEC_ID_VP9);
            this.frameRecorder.setFormat("webm");

            this.frameRecorder.setFrameRate(this.fps);
            this.frameRecorder.setPixelFormat(avutil.AV_PIX_FMT_YUV420P);

            this.frameRecorder.start();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public BufferedImage getScreenshot() {
        return this.cameraScreenshot.getCurrentImage();
    }

    @Override
    public void interrupt() {
        try {
            this.grabFrames = false;
            this.subThread.interrupt();

            if (this.frameRecorder != null)
                this.frameRecorder.stop();

            this.frameGrabber.flush();
            this.frameGrabber.close();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

```

**Comment by saudet** [comment]
What are the warnings and error messages that it displays on the console?

**Comment by MarcWoodyard** [comment]
No error messages are displayed to the console.

**Comment by saudet** [comment]
If you do not see anything, make sure FFmpegLogCallback.set() has been called.

**Comment by MarcWoodyard** [comment]
I added the call in the location below, but nothing was displayed to the console. 

```java
 @Override
    public final void start() {
        try {

            FFmpegLogCallback.set()
         
            this.frameGrabber = new IPCameraFrameGrabber(this.streamURL, 3, 3, TimeUnit.SECONDS);

            if (streamURL.contains("rtsp")) {
                this.frameGrabber.setFormat("rtsp");
                this.frameGrabber.setOption("rtsp_transport", "tcp");
            } else if (streamURL.contains("mjpg"))
                this.frameGrabber.setFormat("mjpeg");
```

**Comment by saudet** [comment]
Try to use FFmpegFrameGrabber instead.

**Comment by MarcWoodyard** [comment]
It still didn't work, here's what was displayed in the console.

```sh
Info: Input #0, mjpeg, from 'http://[hostname]/mjpg/video.mjpg':

Info:   Duration: 
Info: N/A
Info: , bitrate: 
Info: N/A
Info: 

Info:     Stream #0:0
Info: : Video: mjpeg (Baseline), yuvj420p(pc, bt470bg/unknown/unknown), 1280x720 [SAR 1:1 DAR 16:9]
Info: , 
Info: 20 tbr, 
Info: 1200k tbn, 
Info: 20 tbc
Info: 

Warning: [swscaler @ 000002327aa91900] deprecated pixel format used, make sure you did set range correctly
Warning: [swscaler @ 000002327aa91900] deprecated pixel format used, make sure you did set range correctly
Warning: [swscaler @ 000002327aa91900] deprecated pixel format used, make sure you did set range correctly
Warning: [swscaler @ 000002327aa91900] deprecated pixel format used, make sure you did set range correctly
Warning: [swscaler @ 000002327aa91900] deprecated pixel format used, make sure you did set range correctly
Warning: [swscaler @ 000002327aa91900] deprecated pixel format used, make sure you did set range correctly
Warning: [swscaler @ 000002327aa91900] deprecated pixel format used, make sure you did set range correctly
Warning: [swscaler @ 000002327aa91900] deprecated pixel format used, make sure you did set range correctly
Info: Input #0, mjpeg, from 'http://[hostname]/mjpg/video.mjpg':

Info:   Duration: 
Info: N/A
Info: , bitrate: 
Info: N/A
Info: 

Info:     Stream #0:0
Info: : Video: mjpeg (Baseline), yuvj420p(pc, bt470bg/unknown/unknown), 1280x720 [SAR 1:1 DAR 16:9]
Info: , 
Info: 20 tbr, 
Info: 1200k tbn, 
Info: 20 tbc
Info: 

Warning: [swscaler @ 000002327c6bf340] deprecated pixel format used, make sure you did set range correctly
Warning: [swscaler @ 000002327aa91900] deprecated pixel format used, make sure you did set range correctly
Warning: [swscaler @ 000002327aa91900] deprecated pixel format used, make sure you did set range correctly
Warning: [swscaler @ 000002327aa91900] deprecated pixel format used, make sure you did set range correctly
```

**Comment by saudet** [comment]
You'll need to use FFmpegFrameRecorder to see the warnings and errors from that one too.

**Comment by MarcWoodyard** [comment]
Here's the code I used, I changed both the frame grabber and recorder to use FFMPEG.

```java
public class JavaCVStreamRecorder extends Thread {

    private static DateTimeFormatter fileFormatter = DateTimeFormatter.ofPattern("MM-dd-yyyy hh-mm-ss", Locale.ENGLISH);
    private static DateTimeFormatter directoryFormatter = DateTimeFormatter.ofPattern("MM-dd-yyyy", Locale.ENGLISH);
    private FFmpegFrameGrabber frameGrabber;
    private FFmpegFrameRecorder frameRecorder;
    private CameraScreenshot cameraScreenshot = new CameraScreenshot();
    private Thread subThread;
    private String cameraName;
    private String streamURL;
    private int width;
    private int height;
    private LocalDateTime start;
    private LocalDateTime end;
    private boolean recordingEnabled;
    private boolean grabFrames = true;
    private int fps = 20;

    public JavaCVStreamRecorder(int id, String cameraName, String streamURL, int width, int height) {
        this.cameraName = cameraName;
        this.streamURL = streamURL;
        this.width = width;
        this.height = height;
        this.recordingEnabled = CameraManager.getCamera(id).getRecordingEnabled();
    }

    @Override
    public final void start() {
        try {

            FFmpegLogCallback.set();

            this.frameGrabber = new FFmpegFrameGrabber(this.streamURL);
            //this.frameGrabber = new IPCameraFrameGrabber(this.streamURL, 3, 3, TimeUnit.SECONDS);

            if (streamURL.contains("rtsp")) {
                this.frameGrabber.setFormat("rtsp");
                this.frameGrabber.setOption("rtsp_transport", "tcp");
            } else if (streamURL.contains("mjpg"))
                this.frameGrabber.setFormat("mjpeg");

            // Enable hardware acceleration (https://github.com/bytedeco/javacpp-presets/issues/708)
            this.frameGrabber.setOption("hwaccel", "opencl");

            this.frameGrabber.setFrameRate(this.fps);

            this.frameGrabber.start();

            // We're setting the resolution after we start because we need to grab a frame so we can
            // detect the frame's resolution.

            // If width & height are not zero, then it means that we set a specific resolution
            // (video stills for web interface), if they are 0, then it means that we want the
            // full resolution.
            if (this.width != 0 || this.height != 0) {
                this.frameGrabber.setImageWidth(this.width);
                this.frameGrabber.setImageHeight(this.height);
            } else {
                Frame frame = this.frameGrabber.grabFrame();
                this.frameGrabber.setImageWidth(frame.imageWidth);
                this.frameGrabber.setImageHeight(frame.imageHeight);
            }

            if (this.recordingEnabled)
                this.createNewRecording();

            this.subThread = new Thread(() -> {
                Frame frame = null;
                long startTime = 0;
                long videoTS = 0;

                while (grabFrames) {
                    try {
                        if (this.recordingEnabled && LocalDateTime.now().isAfter(this.end)) {
                            this.frameRecorder.stop();
                            this.createNewRecording();
                        }

                        // Prevents Socket Closed exception if thread is .interrupt() is
                        // called while inside of the while loop.
                        try {
                            frame = this.frameGrabber.grab();
                        } catch (Exception e) {
                            if (!e.getMessage().contains("Socket closed"))
                                e.printStackTrace();
                        }

                        if (frame != null) {
                            this.cameraScreenshot.updateImage(frame);

                            if (this.recordingEnabled) {

                                if (startTime == 0)
                                    startTime = System.currentTimeMillis();


                                // Create timestamp for this frame
                                videoTS = 1000 * (System.currentTimeMillis() - startTime);

                                // Check for AV drift
                                if (videoTS > this.frameRecorder.getTimestamp()) {
                                    /*
                                    System.out.println(
                                            "Lip-flap correction: "
                                                    + videoTS + " : "
                                                    + this.frameRecorder.getTimestamp() + " -> "
                                                    + (videoTS - this.frameRecorder.getTimestamp()));
                                     */

                                    // We tell the recorder to write this frame at this timestamp
                                    this.frameRecorder.setTimestamp(videoTS);
                                }

                                this.frameRecorder.record(frame);
                                Thread.sleep((long) (1000 / this.fps));
                            }

                        }
                    } catch (NullPointerException ignored) {
                    } catch (Exception e) {
                        e.printStackTrace();
                    }
                }
            });

            this.subThread.start();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private void createNewRecording() {
        try {

            // TODO Change this to a timer. If time changes, it will create a large file.
            this.start = LocalDateTime.now();
            this.end = LocalDateTime.now().plusMinutes(2);

            File recordingDir;

            do {
                recordingDir = new File(ProgramSettings.getRecordingDirectory() +
                        File.separator +
                        this.cameraName +
                        File.separator +
                        directoryFormatter.format(this.start));

                if (!recordingDir.exists()) {
                    // If we can't create the recording dir specified in ProgramSettings(), then revert to the default location.
                    if (!recordingDir.mkdirs()) {
                        ProgramSettings.setRecordingDirectory(ProgramSettings.getDefaultRecordingDir());
                        ProgramSettings.save();
                    }
                }
            } while (!recordingDir.exists());

            String videoFileLocation = recordingDir.getAbsoluteFile() +
                    File.separator +
                    cameraName + " [" + fileFormatter.format(this.start) + "] - [" + fileFormatter.format(this.end) + "].webm";

            System.out.println(this.frameGrabber.getImageHeight());

            this.frameRecorder = new FFmpegFrameRecorder(
                    videoFileLocation,
                    this.frameGrabber.getImageWidth(),
                    this.frameGrabber.getImageHeight());

            this.frameRecorder.setInterleaved(true);

            // Decrease "startup" latency in FFMPEG (https://trac.ffmpeg.org/wiki/StreamingGuide)
            this.frameRecorder.setVideoOption("tune", "zerolatency");

            // Key frame interval, in our case every 2 seconds -> FPS * 2 = GOP length
            this.frameRecorder.setGopSize(this.fps * 2);

            // Tradeoff between quality and encode speed possible values are ultrafast, superfast, veryfast, faster, fast,
            // medium, slow, slower, veryslow, ultrafast offers us the least amount of compression (lower encoder CPU) at the
            // cost of a larger stream size at the other end, veryslow provides the best compression (high encoder CPU) while
            // lowering the stream size (https://trac.ffmpeg.org/wiki/Encode/H.264)
            this.frameRecorder.setVideoOption("preset", "ultrafast");

            // Constant Rate Factor (https://trac.ffmpeg.org/wiki/Encode/H.264)
            this.frameRecorder.setVideoOption("crf", "28");

            this.frameRecorder.setVideoCodec(avcodec.AV_CODEC_ID_VP9);
            this.frameRecorder.setFormat("webm");

            this.frameRecorder.setFrameRate(this.fps);
            this.frameRecorder.setPixelFormat(avutil.AV_PIX_FMT_YUV420P);

            this.frameRecorder.start();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public BufferedImage getScreenshot() {
        return this.cameraScreenshot.getCurrentImage();
    }

    @Override
    public void interrupt() {
        try {
            this.grabFrames = false;
            this.subThread.interrupt();

            if (this.frameRecorder != null)
                this.frameRecorder.stop();

            this.frameGrabber.flush();
            this.frameGrabber.close();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

}

```
Here's the error message that was displayed in the console.

```sh
org.bytedeco.javacv.FrameRecorder$Exception: avcodec_open2() error -22: Could not open video codec.
	at org.bytedeco.javacv.FFmpegFrameRecorder.startUnsafe(FFmpegFrameRecorder.java:743)
	at org.bytedeco.javacv.FFmpegFrameRecorder.start(FFmpegFrameRecorder.java:401)
	at network.edgeroute.camera.threads.recording.JavaCVStreamRecorder.createNewRecording(JavaCVStreamRecorder.java:208)
	at network.edgeroute.camera.threads.recording.JavaCVStreamRecorder.start(JavaCVStreamRecorder.java:80)
	at network.edgeroute.camera.ThreadManager.startRecorder(ThreadManager.java:60)
	at network.edgeroute.camera.Camera.startRecording(Camera.java:165)
	at network.edgeroute.camera.CameraManager.startAllServerThreads(CameraManager.java:113)
	at network.edgeroute.Launcher.lambda$main$6(Launcher.java:71)
	at java.base/java.lang.Thread.run(Thread.java:832)
```

**Comment by saudet** [comment]
> If you do not see anything, make sure FFmpegLogCallback.set() has been called.

**Comment by MarcWoodyard** [comment]
It's called in the code above.

```java
@Override
    public final void start() {
        try {

            FFmpegLogCallback.set();

            this.frameGrabber = new FFmpegFrameGrabber(this.streamURL);
            //this.frameGrabber = new IPCameraFrameGrabber(this.streamURL, 3, 3, TimeUnit.SECONDS);

            if (streamURL.contains("rtsp")) {
                this.frameGrabber.setFormat("rtsp");
                this.frameGrabber.setOption("rtsp_transport", "tcp");
            } else if (streamURL.contains("mjpg"))
                this.frameGrabber.setFormat("mjpeg");
```

**Comment by saudet** [comment]
You'll need to call it before that.

**Comment by MarcWoodyard** [comment]
I put <code>FFmpegLogCallback.set()</code> at the begining of the program and had the frame output the dimensions to the console.

```sh
Warning: [swscaler @ 000001b3904d7980] deprecated pixel format used, make sure you did set range correctly

Warning: [swscaler @ 000001b3904d7980] deprecated pixel format used, make sure you did set range correctly

Error: [libvpx-vp9 encoder @ 000001b38dcd4a00] [Eval @ 0000005acb1fe130] Undefined constant or missing '(' in 'zerolatency'

Error: [libvpx-vp9 encoder @ 000001b38dcd4a00] Unable to parse option value "zerolatency"

Error: [libvpx-vp9 encoder @ 000001b38dcd4a00] Error setting option tune to value zerolatency.

org.bytedeco.javacv.FrameRecorder$Exception: avcodec_open2() error -22: Could not open video codec.
	at org.bytedeco.javacv.FFmpegFrameRecorder.startUnsafe(FFmpegFrameRecorder.java:743)
	at org.bytedeco.javacv.FFmpegFrameRecorder.start(FFmpegFrameRecorder.java:401)
	at network.edgeroute.camera.threads.recording.JavaCVStreamRecorder.createNewRecording(JavaCVStreamRecorder.java:208)
	at network.edgeroute.camera.threads.recording.JavaCVStreamRecorder.start(JavaCVStreamRecorder.java:77)
	at network.edgeroute.camera.ThreadManager.startRecorder(ThreadManager.java:60)
	at network.edgeroute.camera.Camera.startRecording(Camera.java:165)
	at network.edgeroute.camera.CameraManager.startAllServerThreads(CameraManager.java:113)
	at network.edgeroute.Launcher.lambda$main$6(Launcher.java:73)
	at java.base/java.lang.Thread.run(Thread.java:832)
Warning: [swscaler @ 000001b391fe9a40] deprecated pixel format used, make sure you did set range correctly

org.bytedeco.javacv.FrameRecorder$Exception: No video output stream (Is imageWidth > 0 && imageHeight > 0 and has start() been called?)
	at org.bytedeco.javacv.FFmpegFrameRecorder.recordImage(FFmpegFrameRecorder.java:946)
	at org.bytedeco.javacv.FFmpegFrameRecorder.record(FFmpegFrameRecorder.java:935)
	at org.bytedeco.javacv.FFmpegFrameRecorder.record(FFmpegFrameRecorder.java:928)
	at network.edgeroute.camera.threads.recording.JavaCVStreamRecorder.lambda$start$0(JavaCVStreamRecorder.java:129)
	at java.base/java.lang.Thread.run(Thread.java:832)

Image Width: 1280
Image Height: 720
```

**Comment by saudet** [comment]
Good! I assume you got everything working now?

**Comment by MarcWoodyard** [comment]
Unfortunately not. It's still creating the file, but not recording anything to it.

**Comment by saudet** [comment]
Are you sure stop() is getting called?

**Comment by MarcWoodyard** [comment]
Yes, because when I change this
```java
String videoFileLocation = recordingDir.getAbsoluteFile() +
                    File.separator +
                    cameraName + " [" + fileFormatter.format(this.start) + "] - [" + fileFormatter.format(this.end) + "].mp4";
```

to this 

```java
String videoFileLocation = recordingDir.getAbsoluteFile() +
                    File.separator +
                    cameraName + " [" + fileFormatter.format(this.start) + "] - [" + fileFormatter.format(this.end) + "].webm";
```

It stops working. When I change that line to <code>webm</code>, it creates a file but doesn't record anything to it, It just stays at 0kb. But when I change that line to <code>mp4</code>, it will record the data to a video file, then stop and create a new one after a time limit is reached.

**Comment by saudet** [comment]
So maybe the WebM format doesn't support the VP9 codec: Try another one.

**Comment by MarcWoodyard** [comment]
Ok, I finally figured out what was going on. It seems that there was an issue with my IntelliJ project file because after deleting it and recreating it everything worked... almost. I changed the FrameRecorder to use an FFMPEGFrameRecorder object and I was able to record to WebM, Mkv, mp4 file formats with pretty much any video codec (VP8, VP9, h264, h265 are the ones I tried). So when I change the FrameRecorder to use an OpenCVFrameRecorder object, the issues came back. So it seems that the OpenCVFrameRecorder object only supports recording to .mp4 file formats.

Thanks for the help @saudet!
